### PR TITLE
Add optimizer to push down aggregation through UNION

### LINKED ISSUE #4831: Intermediate aggregation not being applied with union all
This is as simplified as I've been able to get it.

We've got two HDFS-backed tables, in parquet, hive catalog:

```
presto:evs> describe all_events_monthly;
       Column       |  Type   |    Comment    
--------------------+---------+---------------
...
 month_partition    | varchar | Partition Key 
 app_partition      | bigint  | Partition Key 
 ref_type_partition | varchar | Partition Key 
(17 rows)

presto:evs> describe all_install_hash_mapping_monthly;
       Column       |  Type   |    Comment    
--------------------+---------+---------------
...
 month_partition    | varchar | Partition Key 
 app_partition      | bigint  | Partition Key 
 ref_type_partition | varchar | Partition Key 
(11 rows)
```

(removed irrelevant columns)

This query ought to be performing preaggregation, but it isn't, it's transmitting the raw rows (which is far less efficient) to an aggregation remote step:

```
presto:evs> explain (type distributed) select approx_distinct(val) from (select '1' as val from all_events_monthly union all select '2' as val from all_install_hash_mapping_monthly) as "tab";
                                                                                                                                                                                                   Query Plan                                
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct]                                                                                                                                                                                                        
     - Output[_col0] => [approx_distinct:bigint]                                                                                                                                                                                             
             _col0 := approx_distinct                                                                                                                                                                                                        
         - Aggregate => [approx_distinct:bigint]                                                                                                                                                                                             
                 approx_distinct := "approx_distinct"("expr_11")                                                                                                                                                                             
             - RemoteSource[1,2] => [expr_11:varchar]                                                                                                                                                                                        
                                                                                                                                                                                                                                             
 Fragment 1 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [expr]                                                                                                                                                                                                                   
     - Project => [expr:varchar]                                                                                                                                                                                                             
             expr := '1'                                                                                                                                                                                                                     
         - TableScan[hive:hive:evs:all_events_monthly, originalConstraint = true] => []                                                                                                                                                    
                 LAYOUT: hive                                                                                                                                                                                                                
                 HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                    
                     :: [[1, 1592]]                                                                                                                                                                                                          
                 HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                               
                     :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                             
                 HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                            
                     :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [google_ad
                                                                                                                                                                                                                                             
 Fragment 2 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [expr_9]                                                                                                                                                                                                                 
     - Project => [expr_9:varchar]                                                                                                                                                                                                           
             expr_9 := '2'                                                                                                                                                                                                                   
         - TableScan[hive:hive:evs:all_install_hash_mapping_monthly, originalConstraint = true] => []                                                                                                                                      
                 LAYOUT: hive                                                                                                                                                                                                                
                 HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                    
                     :: [[1, 1592]]                                                                                                                                                                                                          
                 HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                               
                     :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                             
                 HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                            
                     :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [google_ad
                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                             
(1 row)
```

This is Presto 0.136 running in AWS (EMR), it used to perform the preaggregation at least in 0.110, possibly also in later versions. (**edit**: no, it doesn't work with 0.110)

In fact, removing the union all (with a simpler query of course), it does perform the preaggregation:

```
presto:evs> explain (type distributed) select approx_distinct(val) from (select '2' as val from all_install_hash_mapping_monthly) as "tab";
                                                                                                                                                                                                     Query Plan                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct]                                                                                                                                                                                                        
     - Output[_col0] => [approx_distinct:bigint]                                                                                                                                                                                             
             _col0 := approx_distinct                                                                                                                                                                                                        
         - Aggregate(FINAL) => [approx_distinct:bigint]                                                                                                                                                                                      
                 approx_distinct := "approx_distinct"("approx_distinct_6")                                                                                                                                                                   
             - RemoteSource[1] => [approx_distinct_6:HyperLogLog]                                                                                                                                                                            
                                                                                                                                                                                                                                             
 Fragment 1 [SOURCE]                                                                                                                                                                                                                         
     Output layout: [approx_distinct_6]                                                                                                                                                                                                      
     - Aggregate(PARTIAL) => [approx_distinct_6:HyperLogLog]                                                                                                                                                                                 
             approx_distinct_6 := "approx_distinct"("expr")                                                                                                                                                                                  
         - Project => [expr:varchar]                                                                                                                                                                                                         
                 expr := '2'                                                                                                                                                                                                                 
             - TableScan[hive:hive:evs:all_install_hash_mapping_monthly, originalConstraint = true] => []                                                                                                                                  
                     LAYOUT: hive                                                                                                                                                                                                            
                     HiveColumnHandle{clientId=hive, name=app_partition, hiveType=int, hiveColumnIndex=-1, partitionKey=true}                                                                                                                
                         :: [[1, 1592]]                                                                                                                                                                                                      
                     HiveColumnHandle{clientId=hive, name=month_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                           
                         :: [[201512], [201601], [201602], [201603]]                                                                                                                                                                         
                     HiveColumnHandle{clientId=hive, name=ref_type_partition, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                        
                         :: [NULL, [android_id], [android_id_md5], [android_id_md5sha1], [android_id_sha1], [apple_ifa], [apple_ifa_md5], [apple_ifa_md5sha1], [apple_ifa_sha1], [google_advertising_id], [google_advertising_id_md5], [googl
                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                             
(1 row)
```

I haven't tested the latest versions, but I've checked the release notes and I haven't seen anything like this mentioned.

Seems like a regression. Applying the partial aggregation step should be perfectly safe with UNION ALL.

### PR DESCRIPTION
**Pull request by geraint0923**
Fixed #4831

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class PushAggregationThroughUnion
+        extends PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public PushAggregationThroughUnion(Metadata metadata)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java]: If this class only needs the FunctionRegistry, pass that directly.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 2]
```diff
@@ -303,30 +304,30 @@ public PlanWithProperties visitAggregation(AggregationNode node, Context context
                 if (child.getProperties().isNodePartitionedOn(partitioningRequirement)) {
                     return rebaseAndDeriveProperties(node, child);
                 }
-                else {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: This cleanup is unrelated. Put it in a separate commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 3]
```diff
@@ -303,30 +304,30 @@ public PlanWithProperties visitAggregation(AggregationNode node, Context context
...
+                        child.getProperties());
+                return rebaseAndDeriveProperties(node, child);
             }
         }
 
         private PlanWithProperties splitAggregation(AggregationNode node, PlanWithProperties newChild, Function<PlanNode, PlanNode> exchanger)
         {
+            // only SINGLE aggregation could be split
+            if (!node.getStep().equals(SINGLE)) {
+                PlanNode resultNode = exchanger == null ? node : exchanger.apply(node);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Rewrite it as follows for better readability:

```java
PlanNode result = node;
if (exchanged != null) {
    result = exchanger.apply(result);
}
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 4]
```diff
@@ -303,30 +304,30 @@ public PlanWithProperties visitAggregation(AggregationNode node, Context context
...
+                return withDerivedProperties(node, newChild.getProperties());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Did you mean `resultNode` instead of `node`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            boolean decomposable = node.getFunctions()
+                    .values().stream()
+                    .map(functionRegistry::getAggregateFunctionImplementation)
+                    .allMatch(InternalAggregationFunction::isDecomposable);
+
+            // Only decomposable aggregations are allowed to perform this optimization
+            if (!decomposable) {
+                return node;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java]: This should be `context.defaultRewrite(node)` so that it walks children recursively.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            }
+
+            PlanNode sourceNode = context.rewrite(node.getSource());
+
+            // For now we don't apply this optimization for aggregation(distinct column),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java]: No need to mention MarkDistinct. This optimization is completely unrelated to that (it's only about Aggregation over Union)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 4 above)
Comment (geraint0923) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Yeah, I mean resultNode here. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        private final SymbolAllocator symbolAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughUnion.java]: The code in this method is too entangled and hard to read. I'd structure it into more clearly-defined sections:

1. For each union input, create a partial aggregation by rewriting the aggregation functions according to the union input-to-output mapping. The body of this loop can be placed in a different method.
2. Allocate names for the outputs of the union.
3. Create a union node on top of the aggregations, mapping each input to the corresponding output from 2)
4. Create the final aggregation on top of the union.

**Comment by geraint0923** [comment]
Addressed the comments. Try my best to put complicated code into separated methods.

**Comment by martint** [comment]
@erichwang, can you review?

**Comment by erichwang** [comment]
So I know that this has already gone through a few rounds of review, but after reviewing this, I think it might actually be cleaner here to isolate aggregation related operations a little more. For example, I think we should extract the aggregation splitting into its own optimizer that runs after AddExchanges. In that case, AddExchanges would only produce the plan with Agg singles. And an optimizer would run afterwards to split the Aggs into partials and move the partials down. This would serve several goals:
1) Moving logical agg splitting into one class
2) Simplifying the complexity of AddExchanges (which is both good for understandability and in the interest of the new planner shape)
3) Allowing further optimizations for partial pushdown and maybe tree-based partial aggregations.

@martint and @geraint0923, what do you guys think?

**Comment by geraint0923** [comment]
Superseded by #5981
